File: src/main/java/gargoyle/ct/convert/impl/SerializableConverter.java
package gargoyle.ct.convert.impl;

import gargoyle.ct.convert.Converter;
import gargoyle.ct.util.CTSerializationUtil;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.Serializable;

public class SerializableConverter<T extends Serializable> implements Converter<T> {
    private final BytesConverter converter = new BytesConverter();

    @Override
    public String format(T data) {
        try {
            return converter.format(CTSerializationUtil.serialize(data));
        } catch (IOException ex) {
            throw new IllegalArgumentException(ex);
        }
    }

    @NotNull
    @Override
    public T parse(@NotNull String data) {
        try {
            return CTSerializationUtil.deserialize(converter.parse(data));
        } catch (IOException ex) {
            throw new IllegalArgumentException(data, ex);
        }
    }
}

File: src/main/kotlin/gargoyle/ct/convert/impl/SerializableConverter.kt
package gargoyle.ct.convert.impl

import gargoyle.ct.convert.Converter
import gargoyle.ct.util.CTSerializationUtil
import java.io.IOException

class SerializableConverter<T : Any> : Converter<T> {
    private val converter = BytesConverter()
    override fun format(data: T): String {
        return try {
            converter.format(CTSerializationUtil.serialize(data))
        } catch (ex: IOException) {
            throw IllegalArgumentException(ex)
        }
    }

    override fun parse(data: String): T {
        return try {
            CTSerializationUtil.deserialize(converter.parse(data))
        } catch (ex: IOException) {
            throw IllegalArgumentException(data, ex)
        }
    }
}
